### ISSUE TITLE
Nullable field is getting ALTER-ed every restart

### ISSUE THREAD

**Issue by jazzfog** [typeorm/typeorm#6459]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

- Defined a nullable field with default null value:

```
  @Column('varchar', { length: 10_000, default: null })
  error!: string;
```
- Restart your app dev server so TypeORM resync the model (do it several times).

Expected result: The field created once and never updated again
Actual result: TypeORM alters the field every restart:

```
query: SELECT VERSION() AS `version`
query: SELECT * FROM `INFORMATION_SCHEMA`.`COLUMNS` WHERE `TABLE_SCHEMA` = 'tmdb_service' AND `TABLE_NAME` = 'typeorm_metadata'
query: ALTER TABLE `task` CHANGE `error` `error` varchar(10000) NULL DEFAULT NULL
query: COMMIT
```

Tested on both MariaDB and MySQL

The bug is **not** present when the field is given a non-`null` default value (e.g. empty string)

**Comment by PadraigGalvin** [comment]
Try using `nullable` instead of setting an empty default.

```js
  @Column('varchar', { length: 10_000, nullable: true })
  error!: string;
```

**Comment by jazzfog** [comment]
@PadraigGalvin,

- It does not help (TypeORM still issues the ALTER instruction every restart)
- This will lead to required value for `error` when you try to save it. It should be optional.

**Comment by PadraigGalvin** [comment]
@jazzfog using `nullable` instead of `default` fixed the issue for me and the `nullable` values are optional. If they are required for you it might be because you are using a [non-null assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

In any case it doesn't really make sense to use `default` if you are not setting a default value.

**Comment by jazzfog** [comment]
You can't fix this with just removing default value, `default` and `nullable` are not interchangeable, they are different things.

The problem has nothing to do with TypeScript.
Was you able to save such record? What value did the record in the DB get? My MySql server complains that nullable column is not getting a value.

> In any case it doesn't really make sense to use default if you are not setting a default value.

You're wrong, I am using a default value, it is `null`. It is not worse than any other default value.

I do not think I am doing anything wrong, _there is no way I should be able to define a model that leads to an `ALTER` every reload_, it is clearly a bug. And it might be a symptom of a bigger issue, like TypeORM table structure extractor is not extracting is not doing it correctly.

**Comment by PadraigGalvin** [comment]
Yeah, there seems to be an issue with comparing the defined default and the current default if they are both `null`.

Glad the workaround helped!

**Comment by AlexMesser** [comment]
This was fixed in #7356 and released in `0.2.32`. Please update your TypeORM version and reopen this issue if the error persists.

**Comment by thefat32** [comment]
@AlexMesser actually this was "unfixed" for mysql in #7356. The bug that #7356 was reported only for postgresql and MysqlDriver was modified too. Already submited a fix in #7580

**Comment by thefat32** [comment]
Actually the pull request has a lot of changes not related with the reported bug. Wonder how it passed the reviews.

**Comment by thefat32** [comment]
For mysql/mariadb should return 'NULL' for normalized 'defaultValue: null'. Also MySql/MariaDB sets default value automatically to null if the column is marked as nullable so we should return also 'NULL' in normalize function when default value is not set and column is marked as nullable

**Comment by thefat32** [comment]
For anyone with this problem. You can workaround this in mysql/mariadb using version `0.2.31` and always use `default` property when using `nullable: true`, also when you want default to be `null`

```ts
{ nullable: true, default: null }
```

This will be fixed when #7580 gets merged
